Alternative docs (#52)
fox-it/dissect-docs: docs/source/target-query.rst

> target-query

Reviewer: I'm missing the target-query -s parameter to convert records to text. Really usefull if you just want to quickly grep through things and if you don't want to deal with rdump selectors.

Author: Thanks, added.

------------

> The **target-query** tool is one of the most prominent tools of Dissect. It allows you to

Reviewer: ```suggestion
The **target-query** tool is one of the most prominent and used tools of Dissect. It allows you to
```

Author: I think this becomes it verbose.

**query a forensic image**
to extract useful information from it. The basic usage format is:

.. code-block:: console

    $ target-query <TARGET> -f <FUNCTION_NAME>

Here <TARGET> is the file that contains the forensic image and <FUNCTION_NAME> is the name

> of the function you wish to apply.

Reviewer: Is it a function or a plugin? In the docs I see both being used.

Reviewer: ```suggestion
.. code-block:: console

    $ target-query <TARGET> -f <FUNCTION_NAME>

Here <TARGET> is the file you wish to query. In most cases this is a forensic image. 
<FUNCTION_NAME> is the name of the function you wish to apply.
```

> Loading

Reviewer: I would maybe not describe this functionality first after the general usage guide of target-query. I think the reader would be way more interested in figuring out which functions/plugins there are.

Author: That's already discussed on the front page. I think some of your comments may stem from the fact that github does not present the documentation in the correct order. Have you also seen the HTML output? I know it is annoying to switch between the HTML and github (to comment), but the view on github might be misleading sometimes.

~~~~~~~

> Dissect attempts to determine the file format automatically and attempts to select an appropriate loader.
> If the image is split into multiple files, just feed it the first.

Reviewer: ```suggestion

By default, Dissect attempts to determine the file format automatically and selects the appropriate loader.
If the image is split into multiple files, just feed it the first. (For example with an EWF image, just target the 
.E01 file) 
```

.. note ::

    The full list of loaders is listed at the end of the output of the ``-l`` option.

In case there is no complete image available but just a couple of separate (log/evt/evtx) files, you can use
the LogLoader for example:

.. code-block:: console

    $ target-query data/*.evtx -L log -f evtx

It is also possible to load multiple separate disks, just string them together using the ``+`` character.

.. code-block:: console

    $ target-query disk1+disk2+disk3 -f osinfo
    
For encrypted disks like LUKS you can provide a key through the ``-K`` option to provide a keychain file
or ``-Kv`` to provide a passphrase.

For more information on the ``-K``, ``--keychain-file`` and ``-Kv``, ``--keychain-value`` arguments, please refer to
:doc:`/usage/disk-encryption`.

    
Querying
~~~~~~~~

To get a full list of all functions available in target-query use the ``-l`` option. If you provide a target image
the list will be filtered based on the compatibility with the target.

You can apply multiple functions if you want:

.. code-block:: console

    $ target-query host.img -f runkeys,users
    
.. note ::

    You can only combine functions that have the same output type.

>     If you don't, only the records will be shown.

Reviewer: A person who has never used target-query has no idea that different plugins have different output types. Maybe describe this at the part where you are explaining the -l option.

Author: Agreed, it is also discussed in the output section, so I removed this notice.

    
If you have a lot of functions you wish to apply, wildcards (and other glob-rules) can be used:

.. code-block:: console

    $ target-query host.img -f apps.browser.*.history
    

> You can combine the ``-f`` option with the `dry-run` option (``-n``/``--dry-run``) to see what will be actually executed:

Reviewer: Whilst this parameter is definitely useful. I wouldn't expect someone who is new to dissect to use it. I would say it is more advanced use of target-query.

Author: It gives people a tool to explore the use of wildcards, so I think that is useful. Also, having a `--dry-run` option encourages people to play a bit, which is good. If the make a mistake, they don't have to wait a long time (because they run the wrong plugins) and they do not see a lot of useless errors. So this is a valuable tool for a beginner and a good step towards understanding the tool better.

.. code-block:: console
    
    target-query host.img -n -q -f apps.browser.*.history
    
Excluding functions is possible with the ``-xf`` flag.

Plugins
~~~~~~~

Each function available in **target-query** is actually a plugin.
Creating your own plugins is also possible. With the ``--plugin-path`` flag you can point
Dissect to your own plugin folder:

.. code-block:: console
    
    target-query host.img -f myplugin --plugin-path=/myplugins

> 

Reviewer: This page is under the "target-query basics" category. I wouldn't say developing plugins is basic usage of dissect/target-query

Author: I see what you mean, but this is more about installing 3rd party plugins than developing your own plugins. I adjusted the text accordingly.

target-query
------------

The **target-query** tool is one of the most prominent tools of Dissect. It allows you to
**query a forensic image**
to extract useful information from it. The basic usage format is:

.. code-block:: console

    $ target-query <TARGET> -f <FUNCTION_NAME>

Here, <TARGET> is the file you wish to query. In most cases, this will be a forensic image.
<FUNCTION_NAME> is the name of the function you wish to apply.

.. note ::

    Here are some :doc:`basic examples </index>` of target-query

Loading
~~~~~~~

> 

Author: Ok, added a line.

.. code-block:: console

    $ target-query disk1+disk2+disk3 -f osinfo
    
For encrypted disks like LUKS you can provide a key through the ``-K`` option to provide a keychain file
or ``-Kv`` to provide a passphrase.

For more information on the ``-K``, ``--keychain-file`` and ``-Kv``, ``--keychain-value`` arguments, please refer to
:doc:`/usage/disk-encryption`.

    
Querying
~~~~~~~~

To get a full list of all functions available in target-query use the ``-l`` option. If you provide a target image
the list will be filtered based on the compatibility with the target.

You can apply multiple functions if you want:

.. code-block:: console

    $ target-query host.img -f runkeys,users
    
If you have a lot of functions you wish to apply, wildcards (and other glob-rules) can be used:

.. code-block:: console

    $ target-query host.img -f apps.browser.*.history
    
You can combine the ``-f`` option with the `dry-run` option (``-n``/``--dry-run``) to see what will be actually executed:

.. code-block:: console
    
    target-query host.img -n -q -f apps.browser.*.history
    
Excluding functions is possible with the ``-xf`` flag.

> 

Author: No, -xf is less important
